### PR TITLE
Multithreading

### PR DESCRIPTION
**Pull request by dstephan1**
- SleepHandler for testing threading - config: `path /zzz SleepHandler { seconds 3; }`
- support for multiple threads - config: `threads 8;`
- integration test for multiple threads

Design choices
- simple as possible
- thread pool managed by boost asio
- no connection pool/manager

### CODE REVIEW

File: config_parser.h [hunk 1]
```diff
@@ -25,6 +26,23 @@ class NginxConfig {
 public:
 	std::string ToString(int depth = 0);
 	std::vector<std::shared_ptr<NginxConfigStatement>> statements_;
+
+	// simple generic find method
+	// returns true on success and changes *out, returns false of failure and doesn't touch *out
+	template <typename T> 
+	bool Find(const std::string& key, T* out) const {
```
Comment (dstephan1) [inline config_parser.h]: This needs a test.

File: config_parser.h [hunk 2]
```diff
@@ -25,6 +26,23 @@ class NginxConfig {
...
+		for (auto statement : statements_)
+		{
+			if (statement->tokens_.size() == 2 && statement->tokens_[0] == key)
+			{
+				// lexical cast using streams
+				std::istringstream stream(statement->tokens_[1]);
+				stream >> *out;
```
Comment (dstephan1) [inline config_parser.h]: What happens if a cast fails?

File: integration_tests.sh [hunk 3]
```diff
@@ -59,6 +63,29 @@ else
...
+for i in {1..15} ; do
+	curl localhost:8080/sleep &
+	processes+=($!)
+done
+sleep 2
+
+for i in "${processes[@]}"; do
+	if ps -p $i > /dev/null
+	then
+		echo "$  !!i too slow, server isn't running in multithreaded"
```
Comment (dstephan1) [inline integration_tests.sh]: typo

File: config_parser.h (hunk 1 above)
Comment (dstephan1) [inline config_parser.h]: done

**Comment by kevinazhu** [comment]
I think this all looks fine. I like the idea of the SleepHandler for easy testing. Good work!

**Pull request merged**